### PR TITLE
Fix un-prefixed xml import bad-sad error

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Fixes the Bad sad when doing `import ballerina/lang.'xml`;

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/21833

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -2592,8 +2592,51 @@ public void visit(BLangXMLQName bLangXMLQName) {
...
+            resultType = symTable.semanticError;
+            return;
+        }
+    }
+
+    private BSymbol getXmlNsSymbolFromPkgSymbol(String localname, String prefix, BSymbol xmlnsSymbol, DiagnosticPos pos) {
+        BPackageSymbol pkgSymbol = (BPackageSymbol) xmlnsSymbol;
+        BSymbol constSymbol = symResolver.lookupMemberSymbol(pos, pkgSymbol.scope, env,
+                names.fromString(localname), SymTag.CONSTANT);
+        if (constSymbol != symTable.notFoundSymbol) {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Can invert this condition right?

File: langlib/lang.xml/src/main/ballerina/src/lang.xml/xml.bal [hunk 2]
```diff
@@ -14,8 +14,17 @@
...
 
-const string XML_NAMESPACE_URI = "http://www.w3.org/XML/1998/namespace";
-const string XMLNS_NAMESPACE_URI = "http://www.w3.org/2000/xmlns/";
+# The namespace URI bound to the `xml` prefix.
+public const string XML_NAMESPACE_URI = "http://www.w3.org/XML/1998/namespace";
+# The namespace URI bound to the `xmlns` prefix.
+public const string XMLNS_NAMESPACE_URI = "http://www.w3.org/2000/xmlns/";
+
+# The expanded name of the `xml:space` attribute.
+public const string space = "{http://www.w3.org/XML/1998/namespace}space";
```
Comment (pubudu91) [inline langlib/lang.xml/src/main/ballerina/src/lang.xml/xml.bal]: Any particular reason why these public constants are in lower case?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -2592,8 +2592,51 @@ public void visit(BLangXMLQName bLangXMLQName) {
...
+            if (constantSymbol.literalType.tag != TypeTags.STRING) {
+                dlog.error(pos, DiagnosticCode.INCOMPATIBLE_TYPES,
+                        symTable.stringType, constantSymbol.literalType);
+                resultType = symTable.semanticError;
+                return null;
+            } else {
+                String constVal = (String) constantSymbol.value.value;
+                int s = constVal.indexOf('{');
+                int e = constVal.lastIndexOf('}');
+                if (e > s) {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: I'm assuming this logic is for constants of the form `"{http://www.w3.org/XML/1998/namespace}space"`? If so, is it a possibility to have multiple braces within such a string? In that case, would the result produced by this logic still be correct?

File: langlib/lang.xml/src/main/ballerina/src/lang.xml/xml.bal (hunk 2 above)
Comment (MaryamZi) [reply langlib/lang.xml/src/main/ballerina/src/lang.xml/xml.bal]: I guess these come from the spec.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -2592,8 +2592,51 @@ public void visit(BLangXMLQName bLangXMLQName) {
...
+
+    private BSymbol getXmlNsSymbolFromPkgSymbol(String localname, String prefix, BSymbol xmlnsSymbol, DiagnosticPos pos) {
+        BPackageSymbol pkgSymbol = (BPackageSymbol) xmlnsSymbol;
+        BSymbol constSymbol = symResolver.lookupMemberSymbol(pos, pkgSymbol.scope, env,
+                names.fromString(localname), SymTag.CONSTANT);
+        if (constSymbol != symTable.notFoundSymbol) {
+            BConstantSymbol constantSymbol = (BConstantSymbol) constSymbol;
+            if (constantSymbol.literalType.tag != TypeTags.STRING) {
+                dlog.error(pos, DiagnosticCode.INCOMPATIBLE_TYPES,
+                        symTable.stringType, constantSymbol.literalType);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]:
```suggestion
                dlog.error(pos, DiagnosticCode.INCOMPATIBLE_TYPES, symTable.stringType, constantSymbol.literalType);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 5]
```diff
@@ -2592,8 +2592,51 @@ public void visit(BLangXMLQName bLangXMLQName) {
...
+            xmlnsSymbol = getXmlNsSymbolFromPkgSymbol(bLangXMLQName.localname.value, bLangXMLQName.prefix.value,
+                    xmlnsSymbol, bLangXMLQName.pos);
+        }
+        if (xmlnsSymbol != null && xmlnsSymbol.getKind() == SymbolKind.XMLNS) {
+            bLangXMLQName.namespaceURI = ((BXMLNSSymbol) xmlnsSymbol).namespaceURI;
+            bLangXMLQName.nsSymbol = (BXMLNSSymbol) xmlnsSymbol;
+        } else {
+            resultType = symTable.semanticError;
+            return;
+        }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]:
```suggestion
        if (xmlnsSymbol == null || xmlnsSymbol.getKind() != SymbolKind.XMLNS) {
            resultType = symTable.semanticError;
            return;
        }
        BXMLNSSymbol symbol = (BXMLNSSymbol) xmlnsSymbol;
        bLangXMLQName.namespaceURI = symbol.namespaceURI;
        bLangXMLQName.nsSymbol = symbol;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 6]
```diff
@@ -2592,8 +2592,51 @@ public void visit(BLangXMLQName bLangXMLQName) {
...
+            } else {
+                String constVal = (String) constantSymbol.value.value;
+                int s = constVal.indexOf('{');
+                int e = constVal.lastIndexOf('}');
+                if (e > s) {
+                    String nsURI = constVal.substring(s + 1, e);
+                    xmlnsSymbol = new BXMLNSSymbol(names.fromString(constVal.substring(e)), nsURI,
+                            constantSymbol.pkgID, constantSymbol.owner);
+                    pkgSymbol.isUsed = true;
+                    return xmlnsSymbol;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]:
```suggestion
                    pkgSymbol.isUsed = true;
                    return new BXMLNSSymbol(names.fromString(constVal.substring(e)), nsURI,
                                            constantSymbol.pkgID, constantSymbol.owner);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 7]
```diff
@@ -2592,8 +2592,51 @@ public void visit(BLangXMLQName bLangXMLQName) {
...
+                }
+                dlog.error(pos, DiagnosticCode.INVALID_ATTRIBUTE_REFERENCE, prefix + ":" + localname);
+                resultType = symTable.semanticError;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Do we need to set `resultType` here? Ideally should be done where called right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 3 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Yey, it's URI, not an HTTP URL

**Pull request merged**